docs/source/core/core.rst

```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
 
-Neural Module
--------------
-Neural Modules are building blocks for Models.
-They accept (typed) inputs and return (typed) outputs. *All Neural Modules inherit from ``torch.nn.Module`` and, therefore, compatible with PyTorch ecosystem.* There are 3 types on Neural Modules:
+.. note:: *All Neural Modules inherit from ``torch.nn.Module`` and, therefore, compatible with PyTorch ecosystem.*
+
+There are 3 types on Neural Modules:
 
     - Regular modules
     - Dataset/IterableDataset
     - Losses
 
+Every Neural Module in NeMo must inherit from `nemo.core.classes.module.NeuralModule` class.
+
+.. autoclass:: nemo.core.classes.module.NeuralModule
+
+Every Neural Modules inherits `nemo.core.classes.common.Typing` interface and needs to define neural types of it's inputs and outputs.
+This is done by defining two properties: `input_types` and `output_types`.
+Each property should return an ordered dictionary of "port name"->"port neural type" pairs.
+This is the example from `nemo.collections.asr.modules.conv_asr.ConvASREncoder` class:
+
+.. code-block:: python
+
+    @property
+    def input_types(self):
+        return OrderedDict(
+            {
+                "audio_signal": NeuralType(('B', 'D', 'T'), SpectrogramType()),
+                "length": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    @property
+    def output_types(self):
+        return OrderedDict(
+            {
+                "outputs": NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation()),
+                "encoded_lengths": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    @typecheck()
+    def forward(self, audio_signal, length=None):
+        ...
+
+The code snippet above means that `nemo.collections.asr.modules.conv_asr.ConvASREncoder` expects two arguments:
+    * First one, named `audio_signal` of shape [batch, dimension, time] with elements representing spectrogram values.
+    * Second one, named `length` of shape [batch] with elements representing lengths of corresponding signals.
+It also means that `.foraward(...)` and `__call__(...)` methods each produce two outputs:
+    * First one, of shape [batch, dimension, time] but with elements representing encoded representation (AcousticEncodedRepresentation class)
+    * Second one, of shape [batch], corresponding to their lengths.
+
+.. tip:: It is a good practice to define types and add `@typecheck()` decorator to your `.forward()` method once your module is ready for use by others.
+
+.. note:: The output of `.forward(...)` method will always be of type `torch.Tensor`, so it will work with any other Pytorch code. The type information will be carried with every tensor.
+If tensors without types will be passed to your module it will not fail but types will note be checked. Thus it is recommended to define input/output types for all your modules, starting with
```

Comment:
```suggestion
If tensors without types will be passed to your module it will not fail but types will not be checked. Thus it is recommended to define input/output types for all your modules, starting with
```